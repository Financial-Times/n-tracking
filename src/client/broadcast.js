export function broadcast (name, data, bubbles = true) {
	const rootEl = Element.prototype.isPrototypeOf(this) ? this : document.body;

	// This requires a polyfill for IE
	// <https://caniuse.com/#feat=customevent>
	const event = new CustomEvent(name, {
		bubbles: bubbles,
		cancelable: true,
		detail: data
	});

	rootEl.dispatchEvent(event);
}
