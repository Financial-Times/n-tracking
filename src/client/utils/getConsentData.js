import { getUsPrivacyForTracking } from "@financial-times/privacy-us-privacy";
import { getPersonalisedConsent } from "@financial-times/ads-personalised-consent";

export default async function getConsentData() {
	const consentValues = (await getPersonalisedConsent()).isAllowed();
	const usprivacy = getUsPrivacyForTracking();
	return {
		...consentValues,
		usprivacy,
	};
}
