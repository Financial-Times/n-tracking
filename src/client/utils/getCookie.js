export function getCookie (name) {
	const regex = new RegExp(`${name}=([^;]+)/`);
	const match = document.cookie && document.cookie.match(regex);
	return match ? match[1] : null;
}
