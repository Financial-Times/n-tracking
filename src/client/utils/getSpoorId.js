export function getSpoorId () {
	const match = document.cookie && document.cookie.match(/spoor-id=([^;]+)/);
	return match ? match[1] : null;
}
