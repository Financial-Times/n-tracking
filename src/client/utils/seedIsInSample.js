import seedRandom from 'seed-random';

export function seedIsInSample (seed, percent) {
	const randomNumberGenerator = seedRandom(seed);
	const pseudoRandomNumber = randomNumberGenerator();

	return Math.round(pseudoRandomNumber * 100) <= percent;
}
