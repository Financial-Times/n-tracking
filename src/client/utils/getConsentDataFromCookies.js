import { getCookie } from "./getCookie";
import { getUsPrivacyForTracking } from "@financial-times/privacy-us-privacy";

export default function getConsentDataFromCookies() {
	const usprivacy = getUsPrivacyForTracking();
	const rawConsentCookie = getCookie("FTConsent");

	if (!rawConsentCookie) {
		return {
			behavioral: false,
			demographic: false,
			programmatic: false,
			all: false,
			usprivacy,
		};
	}

	const { behaviouraladsOnsite, demographicadsOnsite, programmaticadsOnsite } = parseFTConsentCookie(rawConsentCookie);

	return {
		behavioral: behaviouraladsOnsite,
		demographic: demographicadsOnsite,
		programmatic: programmaticadsOnsite,
		all: behaviouraladsOnsite && demographicadsOnsite && programmaticadsOnsite,
		usprivacy,
	};
}

function parseFTConsentCookie(value) {
	if (!value || !value.length) {
		return {
			behaviouraladsOnsite: false,
			demographicadsOnsite: false,
			programmaticadsOnsite: false,
		};
	}

	return Object.fromEntries(
		decodeURIComponent(value)
			.split(",")
			.map((item) => item.split(":"))
			.map(([key, value]) => [key, value === "on" ? true : false])
	);
}
