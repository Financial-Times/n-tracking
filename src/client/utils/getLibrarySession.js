import { getCookie } from './getCookie';

export const getLibrarySession = () => getCookie('FTSession_PubLib');
