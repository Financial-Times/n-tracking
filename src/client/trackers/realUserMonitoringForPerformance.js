import {getLCP, getFID, getCLS} from 'web-vitals';
import readyState from 'ready-state';
import { broadcast } from '../broadcast';
import { seedIsInSample } from '../utils/seedIsInSample';
import { getSpoorId } from '../utils/getSpoorId';

// @see "Important metrics to measure" https://web.dev/metrics
// @see https://github.com/GoogleChrome/web-vitals#batch-multiple-reports-together

const queue = new Set();

const samplePercentage = 100;

export const realUserMonitoringForPerformance = () => {

	// Gather metrics for only a cohort of users.
	const spoorId = getSpoorId();
	if (!seedIsInSample(spoorId, samplePercentage)) return;

	//Add the web-vitals stats we want to the queue ready for the page to be unloaded or hidden
	//at which point we will send them
	function addToQueue (metric) {
		queue.add(metric);
	}
	getCLS(addToQueue);
	getFID(addToQueue);
	getLCP(addToQueue);

	let context = {};

	const navigation = performance.getEntriesByType('navigation')[0];

	// Proceed only if the page load event is a "navigate".
	// @see: https://developer.mozilla.org/en-US/docs/Web/API/PerformanceNavigationTiming/type
	if (navigation.type !== 'navigate') return;

	readyState.complete.then(() => {
		// <https://developer.mozilla.org/en-US/docs/Web/API/PerformanceNavigationTiming/domInteractive>
		context.domInteractive = Math.round(navigation.domInteractive);
		// <https://developer.mozilla.org/en-US/docs/Web/API/PerformanceNavigationTiming/domComplete>
		context.domComplete = Math.round(navigation.domComplete);
	});

	// Report all available metrics whenever the page is backgrounded or unloaded.
	addEventListener('visibilitychange', () => {
		if (document.visibilityState === 'hidden') {
			flushQueue();
		}
	});

	// NOTE: Safari does not reliably fire the `visibilitychange` event when the
	// page is being unloaded. If Safari support is needed, you should also flush
	// the queue in the `pagehide` event.
	addEventListener('pagehide', flushQueue);

	function flushQueue () {
		if (queue.size > 0) {

			for (const metric of queue) {
				if(metric.name === 'FID') {
					context.firstInputDelay = Math.round(metric.value);
				} else if (metric.name === 'CLS') {
					//as this can be a very long floating number, am rounding to 4 decimal places
					context.cumulativeLayoutShift = +((metric.value).toFixed(4));
				} else if (metric.name === 'LCP') {
					context.largestContentfulPaint = Math.round(metric.value);
				}
			}

			//at this point, send metrics
			broadcast('oTracking.event', {
				action: 'performance',
				category: 'page',
				context
			});
			queue.clear();
		}
	}

};
