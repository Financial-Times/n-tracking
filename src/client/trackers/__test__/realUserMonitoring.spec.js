
// Mock the ready-state module
jest.mock('ready-state', () => ({
	complete: {
		// We want ready-state to complete immediately
		then: (fn) => fn()
	}
}));

// Mock the utility to check whether the seed is in the sample
jest.mock('../../utils/seedIsInSample', () => ({
	seedIsInSample: jest.fn().mockReturnValue(true)
}));
import {seedIsInSample} from '../../utils/seedIsInSample';

// Mock the utility to get a spoor ID from the cookie
jest.mock('../../utils/getSpoorId', () => ({
	getSpoorId: jest.fn().mockReturnValue('mock-spoor-id')
}));

// Mock the broadcast method which fires an o-tracking event
jest.mock('../../broadcast', () => ({
	broadcast: jest.fn()
}));
import {broadcast} from '../../broadcast';

// Mock Perfume
jest.mock('perfume.js', () => {
	return jest.fn();
});
import Perfume from 'perfume.js';

// Mock global performance metrics
Performance.prototype.getEntriesByType = jest.fn().mockReturnValue([
	{
		type: 'navigate',
		domComplete: 123.45,
		domInteractive: 678.90
	}
]);

// Mock requestIdleCallback so we can capture metrics
// sent by Perfume
window.requestIdleCallback = jest.fn();

import {realUserMonitoringForPerformance} from '../realUserMonitoringForPerformance';

describe('src/client/trackers/realUserMonitoringForPerformance', () => {

	describe('.realUserMonitoringForPerformance()', () => {

		beforeEach(() => {
			realUserMonitoringForPerformance();
		});

		it('creates a Perfume instance with a custom analytics tracker', () => {
			expect(Perfume).toHaveBeenCalledTimes(1);
			const perfumeOptions = Perfume.mock.calls[0][0];
			expect(typeof perfumeOptions).toBe('object');
			expect(typeof perfumeOptions.analyticsTracker).toBe('function');
		});

		describe('analyticsTracker(options)', () => {
			let analyticsTracker;
			let oldConsole;

			beforeEach(() => {
				// We mock the console here so that our library's console.logs don't
				// make it hard to read the test output
				oldConsole = global.console;
				global.console = {log: jest.fn()};
				analyticsTracker = Perfume.mock.calls[0][0].analyticsTracker;
			});

			afterEach(() => {
				global.console = oldConsole;
			});

			describe('when only one metric type is sent', () => {
				beforeEach(() => {
					analyticsTracker({ metricName: 'fid', data: 13.7 });
				});

				it('does not broadcast an o-tracking event', () => {
					expect(broadcast).toHaveBeenCalledTimes(0);
				});
			});

			describe('when all metrics are sent', () => {
				beforeEach(() => {
					analyticsTracker({ metricName: 'fid', data: 13.7 });
					analyticsTracker({ metricName: 'lcp', data: 13.7 });
					analyticsTracker({ metricName: 'ttfb', data: 13.7 });
					analyticsTracker({ metricName: 'fp', data: 13.7 });
					analyticsTracker({ metricName: 'fcp', data: 13.7 });
					analyticsTracker({ metricName: 'cls', data: 13.7 });
					analyticsTracker({ metricName: 'tbt', data: 13.7 });
				});

				it('broadcasts an o-tracking event with the formatted metrics data', () => {
					expect(broadcast).toHaveBeenCalledTimes(1);
					const broadcastArguments = broadcast.mock.calls[0];
					expect(broadcastArguments[0]).toBe('oTracking.event');
					expect(broadcastArguments[1]).toMatchSnapshot();
				});

				describe('when any one of the metrics are sent a second time', () => {
					beforeEach(() => {
						analyticsTracker({ metricName: 'fid', data: 13.7 });
					});

					it('does not broadcast a second o-tracking event', () => {
						expect(broadcast).toHaveBeenCalledTimes(1);
					});
				});
			});

		});

		describe('when the seed is not in the sample', () => {

			beforeEach(() => {
				Perfume.mockReset();
				seedIsInSample.mockReturnValue(false);
				realUserMonitoringForPerformance();
			});

			it('does not create a Perfume instance', () => {
				expect(Perfume).toHaveBeenCalledTimes(0);
			});

		});

		describe('when no "navigation" performance entries are available', () => {

			beforeEach(() => {
				Perfume.mockReset();
				Performance.prototype.getEntriesByType.mockReturnValue([]);
				realUserMonitoringForPerformance();
			});

			it('does not create a Perfume instance', () => {
				expect(Perfume).toHaveBeenCalledTimes(0);
			});

		});

	});
});
