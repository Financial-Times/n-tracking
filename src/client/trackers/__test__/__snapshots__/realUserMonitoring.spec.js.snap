// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/client/trackers/realUserMonitoringForPerformance .realUserMonitoringForPerformance() recordMetric(metric) when all metrics are sent broadcasts an o-tracking event with the formatted metrics data 1`] = `
Object {
  "action": "performance",
  "category": "page",
  "cumulativeLayoutShift": 13.7654,
  "domComplete": 123,
  "domInteractive": 679,
  "firstContentfulPaint": 14,
  "firstInputDelay": 14,
  "largestContentfulPaint": 14,
  "timeToFirstByte": 14,
  "url": "http://localhost/",
}
`;
